src/image-runner.ts

```diff
@@ -0,0 +1,151 @@
+// external
+import * as express from 'express';
+import { ContainerInfo } from 'dockerode';
+
+// internal
+import {
+	getAllImages,
+	findContainer,
+	pullImage,
+	deleteContainer,
+	ContainerName,
+	proxyRequestToContainer,
+	reviveContainer,
+	createContainer,
+} from './api';
+import { config } from './config';
+import { l, ringbuffer } from './logger';
+
+const imagePattern = /^container-(?<container>\w+)\./;
+
+function stripImageHashSubdomainFromHost( host: string ) {
+	return host.replace( imagePattern, '' );
+}
+
+function assembleSubdomainUrlForContainer( req: express.Request, container: ContainerInfo ) {
+	const protocol = req.secure || req.headers.host.indexOf( 'calypso.live' ) > -1 ? 'https' : 'http';
+	const environment = container.Labels[ 'calypsoEnvironment' ];
+
+	const subdomainEnv = environment && environment !== config.envs[ 0 ] ? environment + '-' : '';
+	// The first character is a `/`, skip it
+	const name = container.Names[ 0 ].substring( 1 );
+
+	const newUrl = new URL(
+		`${ protocol }://${ subdomainEnv }container-${ name }.${ stripImageHashSubdomainFromHost(
+			req.headers.host
+		) }`
+	);
+	newUrl.pathname = req.path;
+	for ( let [ key, value ] of Object.entries( req.query ) ) {
+		if ( key === 'hash' || key === 'branch' || key === 'env' || key === 'image' ) {
+			continue;
+		}
+		newUrl.searchParams.set( key, String( value ) );
+	}
+
+	return newUrl.toString();
+}
+
+function getContainerNameFromSubdomain( host: string ) {
+	const match = host.match( imagePattern );
+	if ( ! match ) {
+		return null;
+	}
+
+	return match.groups.container;
+}
+
+/**
+ * Gets an image name from the query string, finds (or creates) a container for that image
+ * and redirects to http://container-<containername>.calypso.live
+ */
+async function loadImage( req: express.Request, res: express.Response ) {
+	const imageName = req.query.image;
+	const environment = req.query.env || config.envs[ 0 ];
+
+	// There is a container for this image/environment. Redirect to http://container-<name>.calypso.live
+	const existingContainer = findContainer( {
+		image: imageName,
+		env: environment,
+	} );
+	if ( existingContainer ) {
+		res.redirect( assembleSubdomainUrlForContainer( req, existingContainer ) );
+		return;
+	}
+
+	// There is no a container for this image, but the image exists in our repo. Create the container and redirect
+	if ( getAllImages().has( imageName ) ) {
+		const container = await createContainer( imageName, environment );
+		res.redirect( assembleSubdomainUrlForContainer( req, container ) );
+		return;
+	}
+
+	// Neither the container nor the image exits. Pull the image, create the container and redirect. If they image is
+	// already being pulled, this will "attach" to the output of the existing pull.
+	// TODO: This poor-man's log may cause problems if the client doesn't understand JavaScript, it will never redirect.
+	res.write( '<!DOCTYPE html><body><pre>' );
```

Comment:
I'm in two minds about this. I can see how this is helpful to humans (pulling an image can take up to 1 minute in my machine, and a log is better than a blank screen), but I can also see how this can break computers (specially the redirect to the subdomain, as it is not an HTTP redirect)

Reply:
Perhaps we could still render a human-readable response but provide something other than just the default 200 response status code (perhaps 204 or 206)? That way we can look for that more specific status in any systems integrating with dserve and handle the situation (i.e. by polling or waiting for the response to complete).